pkg/v1/v1util/zip.go

```diff
@@ -86,33 +86,45 @@ func GzipWriteCloser(w io.WriteCloser) io.WriteCloser {
 	}
 }
 
-// gunzipWriteCloser implements io.WriteCloser
-// It is used to implement GunzipWriteClose.
-type gunzipWriteCloser struct {
-	*bytes.Buffer
-	writer io.WriteCloser
-}
-
-// Close implements io.WriteCloser
-func (gwc *gunzipWriteCloser) Close() error {
-	// TODO(mattmoor): How to avoid buffering this whole thing into memory?
-	gr, err := gzip.NewReader(gwc.Buffer)
-	if err != nil {
-		return err
-	}
-	if _, err := io.Copy(gwc.writer, gr); err != nil {
-		return err
-	}
-	return gwc.writer.Close()
-}
-
 // GunzipWriteCloser returns an io.WriteCloser to which compressed data may be
 // written, and the uncompressed data is then written to the provided
 // io.WriteCloser.
 func GunzipWriteCloser(w io.WriteCloser) (io.WriteCloser, error) {
-	return &gunzipWriteCloser{
-		Buffer: bytes.NewBuffer(nil),
-		writer: w,
+	flushed := make(chan bool, 1)
+
+	pr, pw := io.Pipe()
+	go func() {
+		defer pr.Close()
+
+		gr, err := gzip.NewReader(pr)
+		if err != nil {
+			pw.CloseWithError(err)
+			return
+		}
+		defer gr.Close()
+
+		_, err = io.Copy(w, gr)
+		if err != nil {
```

Comment:
You can just call `CloseWithError(err)` here, if `err == nil` it's equivalent to `Close` (which you should call instead of using the flushed chan).